### DISCUSSION TITLE
Comment API for uploading images/attachemnts for use in comments

### DISCUSSION THREAD

**Discussion by tingleby** [community/community#28219] [Apps, API and Webhooks]
Seemingly there is not an API for uploading images when commenting via the API. Is this something that might likely be added or is the omission intended.

Stack overflow reference, https://stackoverflow.com/questions/41581151/how-to-upload-an-image-to-use-in-issue-comments-via-github-api

**Comment by DenverCoder1** [comment]
Why do you need the images hosted on GitHub's cdn? Can't you just use any image hosting server or API? It would show up the same regardless.

**Reply by tingleby** [reply]
Yes that could be an option but ideally keeping it on GitHub would be preferred especially for Private/Enterprise projects.

**Reply by gm3dmo** [reply]
The GitHub documentation on [attaching files](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/attaching-files) calls out the following:

> To keep sensitive media files private, serve them from a private network or server that requires authentication.

**Reply by gm3dmo** [reply]
We've added a link to this discussion to the internal issue tracking requests for the API to have the ability to upload attachments.

**Reply by LukePrior** [reply]
Is there any update on this?

**Reply by MJLHThomassenHadrian** [reply]
+1, we would also like to have this ability to show svg graph stats in a pr thread.

**Reply by back2root** [reply]
@gm3dmo is there any feedback on internal issue tracking, if this request might be implemented.
To me, it would be a handy feature, being able to attach files to comments generated by automated GitHub actions.

**Reply by gm3dmo** [reply]
I feel like it would be handy too.  But then I also feel we should never do it. There is a bit of GitHub that we never interact with          that has to deal with abuse of our systems. i'm guessing this is why we have never done this feature because of the kind of abuses it would sustain.

**Reply by SOF3** [reply]
> Why do you need the images hosted on GitHub's cdn? Can't you just use any image hosting server or API? It would show up the same regardless.

That implies lock-in to a certain CDN provider, which could be unideal especially when you need corporate approval to use a certain image provider.

**Comment by dwisiswant0** [comment]
Any updates on this?

**Comment by midiakiasat** [comment]
There is no public REST/GraphQL endpoint for uploading comment attachments, and that omission is almost certainly intentional.

When you upload an image through the web UI, the client performs a signed upload to an internal asset pipeline (S3 + CDN), then injects the resulting URL into the comment. That flow relies on:

* Authenticated browser session
* Anti-abuse heuristics
* Rate limiting
* Malware scanning
* Content moderation
* Ephemeral upload tokens

Exposing a generic “upload arbitrary binary for comment embedding” API would dramatically expand abuse surface (malware hosting, spam CDN piggybacking, DMCA liability amplification).

For automation use cases (Actions posting charts, SVGs, artifacts), the supported patterns today are:

1. Host externally (S3, GCS, Azure Blob, internal artifact server).
2. Use GitHub Releases assets.
3. Use GitHub Actions artifacts.
4. Embed base64 (for small SVG) directly in markdown when feasible.

If GitHub were to support this, it would likely need:

* Repo-scoped upload tokens
* Strict size/type limits
* Audit visibility
* Quota enforcement

Until then, the lack of API support appears to be a deliberate abuse-prevention boundary, not an oversight.

**Comment by vipulgupta2048** [comment]
I built [**gitshot**](https://github.com/vipulgupta2048/gitshot) to solve exactly this - posted this image using gitshot 🎉 🎉 
GitShot is a zero-config, agent-ready CLI that uploads images to GitHub issues, PRs, and comments. [Announcement](https://x.com/vipulgupta2048/status/2037139503245320667).
No Playwright, no browser automation, no scripts.

<img width="618" height="474" alt="Image" src="https://github.com/user-attachments/assets/c8ef78ff-b14b-4eee-b4dc-2f61333af74e" />

**Comment by drogers0** [comment]
![banner.png](https://github.com/user-attachments/assets/882af717-797b-4680-aca0-3703c84ab4fb)

I built [`gh-image`](https://github.com/drogers0/gh-image) as a `gh` CLI extension to solve this. It replicates the browser's own upload flow, so images land at `github.com/user-attachments/assets/…`.
The same URLs you get from drag-and-drop. **Images on private repos stay private**, since the URLs inherit repo visibility.

```bash
gh extension install drogers0/gh-image

IMG=$(gh image banner.png --repo github-community/community)
# => ![banner](https://github.com/user-attachments/assets/...

gh api graphql ... --body "$IMG
I built gh-image as a \`gh\` CLI extension to solve this...."
```
^ That's the command I used to post this.

It automatically pulls your browser's session cookie, no tokens or OAuth setup required.

**Reply by SOF3** [reply]
Does this work in CI?

**Reply by drogers0** [reply]
Not out of the box. The real problem would be auth since I am pulling a browser cookie that wouldn't exist in CI.

You could make this work in CI by storing a live GitHub user_session cookie as an Actions secret and having the workflow use it to replay the same browser-authenticated upload flow the tool uses locally. That would likely work until GitHub expires or invalidates the session which could be mitigated with a pre-push hook that checks if it is still valid.

However, that would put a real user login session into CI instead of using a supported automation credential, which would be brittle, harder to rotate safely, and riskier if leaked or misused. 

All that being said: It would be trivial for me to implement a session_token passthrough which would enable you to provide tokens from wherever you want, including Actions secrets that you maintain. I'll create an issue on it and add some readme how-to's to give people the option. I'll probably implement over the weekend if you want to check back then.

---

[gitshot](https://github.com/vipulgupta2048/gitshot) mentioned above seems to work natively with CI and minimal setup if you are okay with the image privacy tradeoffs.

**Comment by erwan-joly** [comment]
Re-raising this with a use case that didn't exist when the thread was opened in 2022: **AI coding agents now do a large and growing share of PR and review work through `gh` and the REST/GraphQL APIs**, and the single biggest gap is that they cannot attach an image to a comment.

Concretely, an agent that just made a UI change can render the result, screenshot it, and describe it in prose — but it cannot show it. Every "here's the before/after," "here's the failing screen," "here's the chart the change produces" has to either be dropped or routed through a third-party image host, which is worse for provenance, permanence, and trust than an attachment that lives on GitHub. The web UI has had drag-and-drop upload for a decade; the API parity gap is now a real workflow blocker rather than a nice-to-have.

This keeps resurfacing from the tooling side too — see cli/cli#12960 ("critical for agentic workflows"), closed as a duplicate of cli/cli#1895, both blocked on exactly the platform endpoint this discussion is asking for.

On the abuse concern raised earlier in this thread (malware/spam/DMCA hosting) — that's legitimate, but it argues for a *scoped* endpoint, not for none:

- Gate it behind a fine-grained PAT / GitHub App permission (e.g. `contents: attachments`), so it's opt-in and revocable, unlike the session-cookie flow the community `gh-image`/`gitshot` tools currently reverse-engineer.
- Restrict accepted MIME types to images (+ maybe PDF/logs), enforce a size cap, and only allow attaching to a comment the token can already write.
- Keep the same virus scanning, rate limits, and takedown path the web uploader already runs — the asset pipeline (`user-attachments`) already exists; this is an authenticated front door to it, not new infrastructure.

That would give agents a first-class, auditable path and let GitHub retire the cookie-replay workarounds that exist precisely because there's no sanctioned option today. Would love an official signal on whether a scoped upload endpoint is on the roadmap.
